rcl/src/rcl/parameter_client.c

```diff
@@ -0,0 +1,301 @@
+  default_options.allocator = rcl_get_default_allocator();
+  default_options.remote_node_name = NULL;
+  return default_options;
+}
+
+rcl_parameter_client_t
+rcl_get_zero_initialized_parameter_client(void)
+{
+  static rcl_parameter_client_t null_client = {0};
+  return null_client;
+}
+
+#define RCL_PARAMETER_INITIALIZE_CLIENT(VERB, SRV_TYPE_NAME, SRV_SUFFIX) \
+  const rosidl_service_type_support_t * VERB ## _ts = ROSIDL_GET_TYPE_SUPPORT( \
+    rcl_interfaces, srv, SRV_TYPE_NAME \
+  ); \
+  const char * VERB ## _service_name = strcat(node_name, SRV_SUFFIX); \
+  ret = rcl_client_init( \
+    &parameter_client->impl->VERB ## _client, \
+    node, \
+    VERB ## _ts, \
+    VERB ## _service_name, \
+    &client_options); \
+  if (ret != RCL_RET_OK) { \
+    goto fail; \
+  } \
+  rcl_interfaces__srv__ ## SRV_TYPE_NAME ## _Request__init(&parameter_client->impl->VERB ## _request); \
+  rcl_interfaces__srv__ ## SRV_TYPE_NAME ## _Response__init(&parameter_client->impl->VERB ## _response); \
+  parameter_client->impl->VERB ## _sequence_number = 0; \
+
+
+rcl_ret_t
+rcl_parameter_client_init(
+  rcl_parameter_client_t * parameter_client,
+  const rcl_node_t * node,
+  const rcl_parameter_client_options_t * options
+)
+{
+  // TODO goto fail
+  RCL_CHECK_ARGUMENT_FOR_NULL(parameter_client, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(node, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(options, RCL_RET_INVALID_ARGUMENT);
+  if (!node->impl) {
+    RCL_SET_ERROR_MSG("invalid node");
+    return RCL_RET_NODE_INVALID;
+  }
+  if (parameter_client->impl) {
+    RCL_SET_ERROR_MSG("client already initialized, or memory was unintialized");
+    return RCL_RET_ALREADY_INIT;
+  }
+  rcl_ret_t ret;
+  rcl_ret_t fail_ret = RCL_RET_ERROR;
+  rcl_client_options_t client_options = rcl_client_get_default_options();
+  client_options.qos = options->qos;
+  client_options.allocator = options->allocator;
+
+  char * node_name = NULL;
+  if (options->remote_node_name != NULL) {
+    size_t len = strlen(options->remote_node_name);
+    node_name = malloc(len);
+    memcpy(node_name, options->remote_node_name, len);
+  } else {
+    const char * node_name_src = rcl_node_get_name(node);
+    memcpy(node_name, node_name_src, strlen(node_name_src));
+  }
+
+  // Initialize all clients in impl storage
+  RCL_PARAMETER_INITIALIZE_CLIENT(get, GetParameters, "__get_parameters");
+  RCL_PARAMETER_INITIALIZE_CLIENT(get_types, GetParameterTypes, "__get_parameter_types");
+  RCL_PARAMETER_INITIALIZE_CLIENT(set, SetParameters, "__set_parameters");
+  RCL_PARAMETER_INITIALIZE_CLIENT(set_atomically, SetParametersAtomically, "__set_parameters_atomically");
+  RCL_PARAMETER_INITIALIZE_CLIENT(list, ListParameters, "__list_parameters");
+
+  const rosidl_message_type_support_t * event_ts = ROSIDL_GET_TYPE_SUPPORT(
+    rcl_interfaces, msg, ParameterEvent);
+  // TODO Is the parameter_event topic name global or node local?
```

Comment:
Summary from meeting, we'll table this discussion as to not hold up this pr. I'd suggest opening a new issue to discuss it.